src/v/pandaproxy/schema_registry/util.h

```diff
@@ -47,4 +55,39 @@ result<schema_definition> make_schema_definition(std::string_view sv) {
       ss::sstring{str_buf.GetString(), str_buf.GetSize()}};
 }
 
+///\brief The first invocation of one_shot::operator()() will invoke func and
+/// wait for it to finish. Concurrent invocatons will also wait.
+///
+/// On success, all waiters will be allowed to continue. Successive invocations
+/// of one_shot::operator()() will return ss::now().
+///
+/// If func fails, waiters will receive the error, and one_shot will be reset.
+/// Successive calls to operator()() will restart the process.
+class one_shot {
+    enum class state { empty, started, available };
+
+public:
+    explicit one_shot(ss::noncopyable_function<ss::future<>()> func)
+      : _func{std::move(func)} {}
+    ss::future<> operator()() {
+        if (likely(_state == state::available)) {
+            return ss::now();
+        }
+        if (std::exchange(_state, state::started) == state::empty) {
+            _prom_started = {};
+            (void)_func().then_wrapped([this](ss::future<> f) {
```

Comment:
Good question. That's currently managed externally with a gate, but perhaps it should be integrated.

Reply:
seems reasonable to document one_shot and say that lifetime has to be managed by the caller. though in that case it might make sense to complete _func in the foreground so one user of one shot would incur the latency of completing it.